vmngclient/api/versions_utils.py

```diff
@@ -0,0 +1,150 @@
+        devices_versions_repository = {}
+        for device in devices_versions_info:
+            device_all_versions = create_dataclass(DeviceSoftwareRepository, device)
+            device_all_versions.installed_versions = [version for version in device_all_versions.available_versions]
+            device_all_versions.installed_versions.append(device_all_versions.current_version)
+            devices_versions_repository[device_all_versions.device_id] = device_all_versions
+        return devices_versions_repository
+
+    def get_image_version(self, software_image: str) -> Union[str, None]:
+        """
+        Method to get proper software image version
+
+        Args:
+            software_image (str): path to software image
+
+        Returns:
+            Union[str, None]: image version or None
+        """
+
+        image_name = PurePath(software_image).name
+        software_images = self.get_all_software_images()
+        for img in software_images:
+            if image_name in img["availableFiles"]:
+                image_version = img["versionName"]
+                return image_version
+        logger.error(f"Software image {image_name} is not in available images")
+        return None
+
+
+class DeviceVersions:
+    """
+    Methods to prepare devices list for payload
+    """
+
+    def __init__(self, repository: RepositoryAPI, device_category: DeviceCategory):
+        self.repository = repository
+        self.device_category = device_category
+
+    def get_device_list_if_in_available(self, version_to_set_up: str, devices: List[Device]) -> list:
+        """
+        Create version key fir every device dict in device list, if requested version
+        is in available versions
+
+        Args:
+            version_to_set_up (str): requested version
+
+        Returns:
+            list : list of devices
+        """
+        devs = [{"deviceId": dev.uuid, "deviceIP": dev.id, "version": None} for dev in devices]
+        all_dev_versions = self.repository.get_devices_versions_repository(self.device_category)
+        for dev in devs:
+            dev_available_versions = all_dev_versions[str(dev["deviceId"])].available_versions
+            for version in dev_available_versions:
+                if version_to_set_up in version:
+                    dev["version"] = version
+                    break
+            if not dev["version"]:
+                logger.error(f"Software version {version_to_set_up} for {dev} is not included in available_versions")
+        return devs
+
+    def get_device_list_if_in_installed(self, version_to_set_up: str, devices: List[Device]) -> list:
+        """
+        Create version key fir every device dict in device list, if requested version
+        is in available versions
+
+        Args:
+            version_to_set_up (str): requested version
+
+        Returns:
+            None
+        """
+        devs = [{"deviceId": dev.uuid, "deviceIP": dev.id} for dev in devices]
```

Comment:
common code can be moved to private method
__get_device_list_in(attr: str)
similairly it was implemented before: using getattr

Reply:
Done, but I used just one underscore, because I'm using it outside DeviceVersions class.

Reply:
where are public methods gone (dedicated for installed/avaiable) ?